app/src/main/java/org/shadowice/flocke/andotp/Activities/AuthenticateActivity.java

```diff
@@ -22,190 +22,284 @@
+        passwordInput = v.findViewById(R.id.passwordEdit);
+        int inputType = (authMethod == AuthMethod.PASSWORD)
+                ? (InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD)
+                : (InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_VARIATION_PASSWORD);
+        passwordInput.setInputType(inputType);
+        passwordInput.setTransformationMethod(new PasswordTransformationMethod());
+        passwordInput.setOnEditorActionListener(this);
+    }
 
-        authMethod = settings.getAuthMethod();
-        password = settings.getAuthCredentials();
+    private void initUnlockViews(View v) {
+        unlockButton = v.findViewById(R.id.buttonUnlock);
+        unlockButton.setOnClickListener(this);
+        unlockProgress = v.findViewById(R.id.unlockProgress);
+    }
 
-        if (password.isEmpty()) {
-            password = settings.getOldCredentials(authMethod);
-            oldPassword = true;
+    private void cancelBackgroundTask() {
+        TaskFragment taskFragment = findTaskFragment();
+        if (taskFragment != null) {
+            taskFragment.task.cancel();
         }
+        setupUiForTaskState(false);
+    }
 
-        if (authMethod == AuthMethod.PASSWORD) {
-            if (password.isEmpty()) {
-                Toast.makeText(this, R.string.auth_toast_password_missing, Toast.LENGTH_LONG).show();
-                finishWithResult(true, null);
-            } else {
-                passwordLayout.setHint(getString(R.string.auth_hint_password));
-                passwordInput.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);
-            }
-        } else if (authMethod == AuthMethod.PIN) {
-            if (password.isEmpty()) {
-                Toast.makeText(this, R.string.auth_toast_pin_missing, Toast.LENGTH_LONG).show();
-                finishWithResult(true, null);
-            } else {
-                passwordLayout.setHint(getString(R.string.auth_hint_pin));
-                passwordInput.setInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_VARIATION_PASSWORD);
+    private class ProcessLifecycleObserver implements DefaultLifecycleObserver {
+        @Override
+        public void onStop(LifecycleOwner owner) {
+            if (settings.getRelockOnBackground()) {
+                cancelBackgroundTask();
             }
-        } else {
-            finishWithResult(true, null);
         }
+    }
 
-        passwordInput.setTransformationMethod(new PasswordTransformationMethod());
-        passwordInput.setOnEditorActionListener(this);
+    @Override
+    public void onResume() {
+        super.onResume();
+        checkBackgroundTask();
+    }
 
-        Button unlockButton = v.findViewById(R.id.buttonUnlock);
-        unlockButton.setOnClickListener(this);
+    private void checkBackgroundTask() {
+        TaskFragment taskFragment = findTaskFragment();
+        if (taskFragment != null && !taskFragment.task.isCanceled()) {
```

Comment:
Could you do something like this here instead (not sure if this is the best way but it works):
```
if (taskFragment != null) {
    if (taskFragment.task.isCanceled()) {
        passwordInput.setText("");
        passwordInput.requestFocus();

        InputMethodManager keyboard = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        keyboard.showSoftInput(passwordInput, 0);
    } else {
        taskFragment.task.setCallback(this::handleResult);
        setupUiForTaskState(true);
    }
}
```

Otherwise the password stays entered when returning from background, which makes the re-lock useless.

And could exclude the `AuthenticateActivity` from being finished in `destroyIfNotMain()` in the `BaseActivity` as well? Otherwise it will show the "Authentication failed, closing" warning after screen off and end the app.

The manual `showSoftInput` seems to be required after screen-off, otherwise it did not show the keyboard again for me.